internal/mocktc/object.go

```diff
@@ -84,6 +94,19 @@ func (object *Object) StartDownload(name string, payload *tcobject.DownloadObjec
 
 /////////////////////////////////////////////////
 
+// FakeS3Object creates a fake object which is assumed to be stored in mocks3
+// at an object of the same name.  It is up to the caller to create the object
+// in mocks3 if necesary.
+func (object *Object) FakeObject(name string) {
+	object.objects[name] = Obj{
+		uploadRequest:  &tcobject.CreateUploadRequest{},
+		uploadFinished: true,
+		onMockS3:       true,
+	}
+}
+
+/////////////////////////////////////////////////
+
```

Comment:
Although not heavily tested and not yet documented, the `GW_TESTS_USE_EXTERNAL_TASKCLUSTER` env var was intended to be used as a means to be able to select whether tests should run against mocks (for speed) or against a real deployment, for integration testing purposes. If tests create objects outside of the API interface, they would not work if run against a real deployment. Since an object can be created via the API, would it be possible for the tests to use that instead, so that the tests could still be run against a real deployment? My thinking is that we could run tests against a real deployment when making a release, to give us further certainty that things function end-to-end (e.g. this may have helped protect against the taskQueueId failures that we introduced recently, that didn't get picked up by the mocks when the task definition structure changed).

If you like, you could keep this convenience method, but in the body wrap the object creation API call rather than updating the underlying map directly, to preserve the capability.

Reply:
The problem with testing against a "real thing" is that it depends on the configuration of that real thing, and requires quite a bit of complexity to establish the conditions for the test -- in this case, creating and claiming a task, creating an artifact on that task, and uploading that artifact.  The result is that the condition-establishment becomes more brittle than the test itself.  In this case, it would depend on configuration of the object service's backends, for example, as well as taskQueues.  It also eliminates the possibility of lots of conditions.  In the example of upload or download support, for example, there's no way to test for upload failures.  And finally, it ties development into the deployment loop: a test won't succeed until the change being tested is running on the TC deployment where CI runs.  That ends up adding weeks to development cycles, as we saw with the client upload/download functionality.  And, it's worth noting, that's not even completely tested, as writing those tests is hard.

In short, I don't think that tests which run against both an upstream Taskcluster deployment and a mock are practical.  Given more engineering bandwidth, maybe we could do some interesting things (perhaps spin up a deployment with minikube for every PR?) but those ideas are immensely far outside our capacity right now.

I appreciate the introduction of a mock, and the use of interfaces to ensure at least some amount of fidelity to the real thing.  But even that is difficult -- I spent a few hours getting the mocks/tests in this PR to work.